RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+
+# Additional classes
+
+1. `MultiTestRunFinalizationCompleteEventArgs` used by IMultiTestRunFinalizationEventsHandler:
+
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.Client
+{
+    [DataContract]
+    public class MultiTestRunFinalizationCompleteEventArgs : EventArgs
+    {
+        /// <summary>
+        /// Default constructor.
+        /// </summary>
+        /// <param name="isCanceled">Specifies whether the finalization is canceled.</param>
+        /// <param name="error">Specifies the error encountered during the execution of the finalization.</param>
+        public MultiTestRunFinalizationCompleteEventArgs(bool isCanceled, Exception error)
+        {
+            this.IsCanceled = isCanceled;
+            this.Error = error;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the finalization is canceled or not.
+        /// </summary>
+        [DataMember]
+        public bool IsCanceled { get; private set; }
+
+        /// <summary>
+        /// Gets the error encountered during the finalization of the test runs. Null if there is no error.
+        /// </summary>
+        [DataMember]
+        public Exception Error { get; private set; }
+
+        /// <summary>
+        /// Get or Sets the Metrics
+        /// </summary>
+        [DataMember]
+        public IDictionary<string, object> Metrics { get; set; }
+    }
+}
+```
+
+2. `MultiTestRunFinalizationProgressEventArgs` used by IMultiTestRunFinalizationEventsHandler:
+
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.Client
+{
+    [DataContract]
+    public class MultiTestRunFinalizationProgressEventArgs : EventArgs
+    {
+        /// <summary>
+        /// Default constructor.
+        /// </summary>
+        /// <param name="currentProcessorIndex">Specifies current processor index.</param>
+        /// <param name="currentProcessorUris">Specifies current processor Uris.</param>
+        /// <param name="currentProcessorProgress">Specifies current processor progress.</param>
+        /// <param name="processorsCount">Specifies the overall number of processors.</param>
+        public MultiTestRunFinalizationProgressEventArgs(long currentProcessorIndex, ICollection<Uri> currentProcessorUris, long currentProcessorProgress, long processorsCount)
+        {
+            CurrentProcessorIndex = currentProcessorIndex;
+            CurrentProcessorUris = currentProcessorUris;
+            CurrentProcessorProgress = currentProcessorProgress;
+            ProcessorsCount = processorsCount;
+        }
+
+        /// <summary>
+        /// Gets a current processor index.
+        /// </summary>
+        [DataMember]
+        public long CurrentProcessorIndex { get; private set; }
```

Comment:
Our code coverage report merger is merging files and removing not needed reports as DevOps is searching for all files with .coverage extension. Because of that we shouldn't run anything in parallel for same uris.

But in case we have two processors which are working on disjoint sets of Uris we can easily use them simultaneously

Reply:
just wondering if a single long will suffice, if we'd ever allow to run merges simultaneously, if that's not the case then having a single long is ok